File: Android/app/src/main/java/com/social/app/chat/SocialLinkManager.kt
package com.social.app.chat

import com.social.app.backend.SupabaseManager
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Gestiona el envío/aceptación de "socials" — equivalente Kotlin de
 * SocialLinkManager.swift. Un social requiere aceptación mutua: el perfil
 * público permite chatear antes, pero el vínculo social siempre necesita
 * los dos síes (tabla `socials`).
 *
 * Antes de esta corrección, Android no tenía NINGUNA implementación de este
 * flujo — a diferencia de iOS (que sí tiene `respond()` cableado en
 * AvisosView, aunque `sendSocial()` tampoco se llama desde ningún sitio
 * todavía en ninguna de las dos plataformas: mandar un social requiere
 * conocer el profile_id real de la persona detectada por UWB, y el motor
 * de proximidad hoy solo intercambia un UUID efímero de sesión, no la
 * identidad real — ver LOOP_STATE.md, pendiente principal).
 */
class SocialLinkManager {

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    @Serializable
    private data class NewSocial(
        @SerialName("requester_id") val requesterId: String,
        @SerialName("addressee_id") val addresseeId: String
    )

    suspend fun sendSocial(requesterId: String, addresseeId: String) {
        try {
            SupabaseManager.client.from("socials").insert(NewSocial(requesterId, addresseeId))
        } catch (e: Exception) {
            _errorMessage.value = "No se pudo enviar el social."
        }
    }

    @Serializable
    private data class SocialRow(
        val id: String,
        @SerialName("requester_id") val requesterId: String,
        @SerialName("addressee_id") val addresseeId: String
    )

    @Serializable
    private data class NewChat(
        @SerialName("user_a_id") val userAId: String,
        @SerialName("user_b_id") val userBId: String
    )

    /** Solo el destinatario puede aceptar (lo aplica la política RLS socials_update). */
    suspend fun respond(socialId: String, accept: Boolean) {
        try {
            SupabaseManager.client.from("socials")
                .update({ set("status", if (accept) "accepted" else "declined") }) {
                    filter { eq("id", socialId) }
                }
            if (accept) createChatIfNeeded(socialId)
        } catch (e: Exception) {
            _errorMessage.value = "No se pudo responder al social."
        }
    }

    private suspend fun createChatIfNeeded(socialId: String) {
        try {
            val social = SupabaseManager.client.from("socials")
                .select { filter { eq("id", socialId) } }
                .decodeSingle<SocialRow>()
            SupabaseManager.client.from("chats")
                .insert(NewChat(social.requesterId, social.addresseeId))
        } catch (e: Exception) {
            // Puede fallar si el chat ya existe (constraint unique) — no es un error de usuario.
        }
    }
}

File: Social/Chat/SocialLinkManager.swift
//
//  SocialLinkManager.swift
//  Social
//
//  Gestiona el envío/aceptación de "socials". Un social requiere aceptación
//  mutua: el perfil público permite chatear antes, pero el vínculo social
//  siempre necesita los dos síes (tabla `socials`, Fase 2).
//

import Foundation

@MainActor
final class SocialLinkManager: ObservableObject {

    @Published var errorMessage: String?

    func sendSocial(from requesterID: UUID, to addresseeID: UUID) async {
        struct NewSocial: Encodable {
            let requester_id: UUID
            let addressee_id: UUID
        }
        do {
            try await SupabaseManager.shared.client
                .from("socials")
                .insert(NewSocial(requester_id: requesterID, addressee_id: addresseeID))
                .execute()
        } catch {
            errorMessage = "No se pudo enviar el social: \(error.localizedDescription)"
        }
    }

    /// Solo el destinatario puede aceptar (lo aplica la política RLS `socials_update`).
    func respond(socialID: UUID, accept: Bool) async {
        do {
            try await SupabaseManager.shared.client
                .from("socials")
                .update(["status": accept ? "accepted" : "declined"])
                .eq("id", value: socialID)
                .execute()

            if accept {
                await createChatIfNeeded(socialID: socialID)
            }
        } catch {
            errorMessage = "No se pudo responder al social: \(error.localizedDescription)"
        }
    }

    /// Al aceptar un social, se asegura de que exista un chat para ambos usuarios.
    private func createChatIfNeeded(socialID: UUID) async {
        do {
            let social: SocialLink = try await SupabaseManager.shared.client
                .from("socials")
                .select()
                .eq("id", value: socialID)
                .single()
                .execute()
                .value
            _ = await getOrCreateChat(social.requesterID, social.addresseeID)
        } catch {
            // No crítico: si falla, simplemente no se crea el chat todavía.
        }
    }

    private struct ChatIDRow: Decodable { let id: UUID }

    /// Devuelve el id del chat entre dos usuarios, creándolo si no existe.
    /// Orden CANÓNICO (uuidString menor primero) -- antes esta función
    /// insertaba siempre en el orden (requester, addressee) del social
    /// concreto que la disparaba. `unique(user_a_id, user_b_id)`
    /// (0001_schema.sql) es SENSIBLE al orden: dos caminos distintos hacia
    /// el mismo par de personas en orden invertido habrían creado dos filas
    /// de chat duplicadas en vez de reutilizar una. Hallazgo real
    /// encontrado al construir esta misma función para "Enviar mensaje" en
    /// el sheet de Avisos (AvisosView.swift) -- necesitaba poder crear un
    /// chat con cualquier persona, no solo tras aceptar un social ya
    /// orientado. Equivalente exacto de getOrCreateChat (Kotlin).
    func getOrCreateChat(_ userIDA: UUID, _ userIDB: UUID) async -> UUID? {
        let (a, b) = userIDA.uuidString < userIDB.uuidString ? (userIDA, userIDB) : (userIDB, userIDA)
        do {
            let existing: [ChatIDRow] = try await SupabaseManager.shared.client
                .from("chats")
                .select("id")
                .eq("user_a_id", value: a)
                .eq("user_b_id", value: b)
                .execute()
                .value
            if let first = existing.first { return first.id }

            struct NewChat: Encodable {
                let user_a_id: UUID
                let user_b_id: UUID
            }
            let inserted: ChatIDRow = try await SupabaseManager.shared.client
                .from("chats")
                .insert(NewChat(user_a_id: a, user_b_id: b))
                .select("id")
                .single()
                .execute()
                .value
            return inserted.id
        } catch {
            return nil
        }
    }
}
